share/man/man4/bio.4
.\"	$NetBSD: bio.4,v 1.1.2.1 2007/05/08 10:45:06 pavel Exp $
.\"	$OpenBSD: bio.4,v 1.19 2006/09/20 22:22:37 jmc Exp $
.\"
.\" Copyright (c) 2002 Niklas Hallqvist
.\" Copyright (c) 2006 Marco Peereboom
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
.\" DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
.\" ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd April 8, 2007
.Dt BIO 4
.Os
.Sh NAME
.Nm bio
.Nd Block IO ioctl tunnel pseudo-device
.Sh SYNOPSIS
.Cd "pseudo-device bio"
.Sh DESCRIPTION
The
.Nm
driver provides userland applications
.Xr ioctl 2
access to devices otherwise not found as
.Pa /dev
nodes.
The
.Pa /dev/bio
device node operates by delegating ioctl
calls to a requested device driver.
Only drivers which have registered with the
.Nm
device can be accessed via this interface.
.Pp
The following device drivers register with
.Nm
for volume management:
.Pp
.Bl -tag -width ciss(4)XX -offset indent -compact
.\" .It Xr ami 4
.\" American Megatrends Inc. MegaRAID PATA/SATA/SCSI RAID controller
.\" .It Xr arc 4
.\" Areca Technology Corporation SATA RAID controller
.\" .It Xr ciss 4
.\" Compaq Smart ARRAY 5/6 SAS/SATA/SCSI RAID controller
.It Xr mfi 4
LSI Logic & Dell MegaRAID SAS RAID controller
.El
.Pp
The following ioctl calls apply to the
.Nm bio
device:
.Bl -tag -width BIOCCAPABILITIES
.It Dv BIOCLOCATE
Locate a named device and give back a cookie to the application
for subsequent ioctl calls.
The cookie is used to tunnel further ioctls to the right device.
.It Dv BIOCINQ
Retrieve number of volumes and physical disks for a specific device.
.It Dv BIOCDISK
Retrieve detailed information for the specified physical disk.
Information returned can include status, size, channel, target, lun,
vendor name, serial number and processor device (ses or safte).
.It Dv BIOCVOL
Retrieve detailed information for the specified volume.
Information returned can include status, size, RAID level, number of disks,
device name association (sd?) and vendor name.
.It Dv BIOCALARM
Control the alarm beeper on the device.
Supported states are: disable alarm, enable alarm, silence alarm, status and
test alarm.
.Pp
Note:  These options might not be supported on all hardware.
.It Dv BIOCBLINK
Blink an LED of the specified physical disk.
Supported blink states are: blink LED, unblink LED and blink alarm LED.
.Pp
Note:  This option is only supported if the disk is governed by ses(4) or
safte(4) and the hardware supports hardware blinking.
.It Dv BIOCSETSTATE
Alter the state of specified physical disk.
Supported states are: create hot-spare, online disk and offline disk.
.El
.Sh FILES
.Bl -tag -width /dev/bio -compact
.It Pa /dev/bio
ioctl tunnel device
.El
.Sh SEE ALSO
.Xr ioctl 2 ,
.Xr bioctl 8
.Sh HISTORY
The
.Nm
driver first appeared in
.Ox 3.2 .
.Sh AUTHORS
.An -nosplit
The
.Nm
driver was written by
.An Niklas Hallqvist Aq niklas@openbsd.org .
The API was written by
.An Marco Peereboom Aq marco@openbsd.org .


sys/dev/bio.c
/*	$NetBSD: bio.c,v 1.1.4.1 2007/05/08 10:45:04 pavel Exp $ */
/*	$OpenBSD: bio.c,v 1.9 2007/03/20 02:35:55 marco Exp $	*/

/*
 * Copyright (c) 2002 Niklas Hallqvist.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/* A device controller ioctl tunnelling device.  */

#include <sys/cdefs.h>
__KERNEL_RCSID(0, "$NetBSD: bio.c,v 1.1.4.1 2007/05/08 10:45:04 pavel Exp $");

#include <sys/param.h>
#include <sys/conf.h>
#include <sys/device.h>
#include <sys/event.h>
#include <sys/ioctl.h>
#include <sys/malloc.h>
#include <sys/queue.h>
#include <sys/systm.h>
#include <sys/proc.h>
#include <sys/kauth.h>

#include <dev/biovar.h>

struct bio_mapping {
	LIST_ENTRY(bio_mapping) bm_link;
	struct device *bm_dev;
	int (*bm_ioctl)(struct device *, u_long, caddr_t);
};

LIST_HEAD(, bio_mapping) bios = LIST_HEAD_INITIALIZER(bios);

void	bioattach(int);
int	bioclose(dev_t, int, int, struct lwp *);
int	bioioctl(dev_t, u_long, caddr_t, int, struct lwp *);
int	bioopen(dev_t, int, int, struct lwp *);

int	bio_delegate_ioctl(struct bio_mapping *, u_long, void *);
struct	bio_mapping *bio_lookup(char *);
int	bio_validate(void *);

const struct cdevsw bio_cdevsw = {
        bioopen, bioclose, noread, nowrite, bioioctl,
        nostop, notty, nopoll, nommap, nokqfilter, 0
};


void
bioattach(int nunits)
{
}

int
bioopen(dev_t dev, int flags, int mode, struct lwp *l)
{
	return (0);
}

int
bioclose(dev_t dev, int flags, int mode, struct lwp *l)
{
	return (0);
}

int
bioioctl(dev_t dev, u_long cmd, caddr_t addr, int flag, struct  lwp *l)
{
	struct bio_locate *locate;
	struct bio_common *common;
	char name[16];
	int error, s;

	switch(cmd) {
	case BIOCLOCATE:
	case BIOCINQ:
	case BIOCDISK:
	case BIOCVOL:
		error = kauth_authorize_device_passthru(l->l_cred, dev,
		    KAUTH_REQ_DEVICE_RAWIO_PASSTHRU_READCONF, addr);
		if (error)
			return (error);
		break;
	case BIOCBLINK:
	case BIOCSETSTATE:
		error = kauth_authorize_device_passthru(l->l_cred, dev,
		    KAUTH_REQ_DEVICE_RAWIO_PASSTHRU_WRITECONF, addr);
		if (error)
			return (error);
		break;
	case BIOCALARM: {
		struct bioc_alarm *alarm = (struct bioc_alarm *)addr;
		switch (alarm->ba_opcode) {
		case BIOC_SADISABLE:
		case BIOC_SAENABLE:
		case BIOC_SASILENCE:
		case BIOC_SATEST:
			error = kauth_authorize_device_passthru(l->l_cred, dev,
			    KAUTH_REQ_DEVICE_RAWIO_PASSTHRU_WRITECONF, addr);
			if (error)
				return (error);
			break;
		case BIOC_GASTATUS:
			error = kauth_authorize_device_passthru(l->l_cred, dev,
			    KAUTH_REQ_DEVICE_RAWIO_PASSTHRU_READCONF, addr);
			if (error)
				return (error);
			break;
		default:
			return EINVAL;
		}
		break;
	}
	default:
		return ENOTTY;
	}

	switch (cmd) {
	case BIOCLOCATE:
		locate = (struct bio_locate *)addr;
		error = copyinstr(locate->bl_name, name, 16, NULL);
		if (error != 0)
			return (error);
		locate->bl_cookie = bio_lookup(name);
		if (locate->bl_cookie == NULL)
			return (ENOENT);
		break;

	default:
		common = (struct bio_common *)addr;
		s = splbio();
		if (!bio_validate(common->bc_cookie)) {
			splx(s);
			return (ENOENT);
		}
		error =  bio_delegate_ioctl(
		    (struct bio_mapping *)common->bc_cookie, cmd, addr);
		splx(s);
		return (error);
	}
	return (0);
}

int
bio_register(struct device *dev, int (*ioctl)(struct device *, u_long, caddr_t))
{
	struct bio_mapping *bm;
	int s;

	MALLOC(bm, struct bio_mapping *, sizeof *bm, M_DEVBUF, M_NOWAIT);
	if (bm == NULL)
		return (ENOMEM);
	bm->bm_dev = dev;
	bm->bm_ioctl = ioctl;
	s = splbio();
	LIST_INSERT_HEAD(&bios, bm, bm_link);
	splx(s);
	return (0);
}

void
bio_unregister(struct device *dev)
{
	struct bio_mapping *bm, *next;
	int s;

	s = splbio();
	for (bm = LIST_FIRST(&bios); bm != NULL; bm = next) {
		next = LIST_NEXT(bm, bm_link);

		if (dev == bm->bm_dev) {
			LIST_REMOVE(bm, bm_link);
			free(bm, M_DEVBUF);
		}
	}
	splx(s);
}

struct bio_mapping *
bio_lookup(char *name)
{
	struct bio_mapping *bm;
	int s;

	s = splbio();
	LIST_FOREACH(bm, &bios, bm_link) {
		if (strcmp(name, bm->bm_dev->dv_xname) == 0) {
			splx(s);
			return (bm);
		}
	}
	splx(s);
	return (NULL);
}

int
bio_validate(void *cookie)
{
	struct bio_mapping *bm;

	LIST_FOREACH(bm, &bios, bm_link) {
		if (bm == cookie) {
			return (1);
		}
	}
	return (0);
}

int
bio_delegate_ioctl(struct bio_mapping *bm, u_long cmd, void *addr)
{
	
	return (bm->bm_ioctl(bm->bm_dev, cmd, addr));
}
